sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/LockRenewSubscriber.java

```diff
@@ -0,0 +1,108 @@
+    private static final OffsetDateTime COMPLETED_RENEW_TIME = OffsetDateTime.now();
+    private static final LockRenewalOperation COMPLETED_ONE = new LockRenewalOperation("lock-no-renew",
+        Duration.ZERO, false, (id) -> Mono.empty(), COMPLETED_RENEW_TIME);
+
+    private final ClientLogger logger = new ClientLogger(LockRenewSubscriber.class);
+
+    private final Function<String, Mono<OffsetDateTime>> onRenewLock;
+    private final Duration maxAutoLockRenewal;
+    private final LockContainer<LockRenewalOperation> messageLockContainer;
+    private final CoreSubscriber<? super ServiceBusReceivedMessage> actual;
+
+    LockRenewSubscriber(CoreSubscriber<? super ServiceBusReceivedMessage> actual, Duration maxAutoLockRenewDuration,
+        LockContainer<LockRenewalOperation> messageLockContainer,
+        Function<String, Mono<OffsetDateTime>> onRenewLock) {
+        this.onRenewLock = Objects.requireNonNull(onRenewLock, "'onRenewLock' cannot be null.");
+        this.actual = actual;
+        this.messageLockContainer = Objects.requireNonNull(messageLockContainer,
+            "'messageLockContainer' cannot be null.");
+        Objects.requireNonNull(maxAutoLockRenewDuration, "'maxAutoLockRenewDuration' cannot be null.");
+        this.maxAutoLockRenewal = maxAutoLockRenewDuration;
+    }
+    /**
+     * On an initial subscription, will take the first work item, and request that amount of work for it.
+     * @param subscription Subscription for upstream.
+     */
+    @Override
+    protected void hookOnSubscribe(Subscription subscription) {
+        Objects.requireNonNull(subscription, "'subscription' cannot be null.");
+        actual.onSubscribe(subscription);
+    }
+
+    /**
+     * When upstream has completed emitting messages.
+     */
+    @Override
+    public void hookOnComplete() {
+        logger.info("Upstream has completed.");
+        actual.onComplete();
+    }
+
+    @Override
+    protected void hookOnError(Throwable throwable) {
+        logger.error("Errors occurred upstream.", throwable);
+        actual.onError(throwable);
+        dispose();
+    }
+
+    @Override
+    protected void hookOnNext(ServiceBusReceivedMessage message) {
+        final String lockToken = message.getLockToken();
+        final OffsetDateTime lockedUntil = message.getLockedUntil();
+        final LockRenewalOperation renewOperation;
+
+        if (!Objects.isNull(lockedUntil) && !Objects.isNull(lockToken)) {
+
+            renewOperation = new LockRenewalOperation(lockToken, maxAutoLockRenewal, false, onRenewLock,
+                lockedUntil, message::setLockedUntil);
+
+            try {
+                messageLockContainer.addOrUpdate(lockToken, OffsetDateTime.now().plus(maxAutoLockRenewal),
+                    renewOperation);
+            } catch (Exception e) {
+                logger.error("Exception occurred while updating lockContainer for token [{}].", lockToken, e);
+                onError(e);
```

Comment:
Wouldn't this call user's onError handler? I thought we were not going to notify the user of any error if auto lock renewal fails.

Reply:
Okay, for now I will not call onError, log it only.